src/bun.js/ipc.zig

```diff
@@ -320,330 +286,589 @@ pub fn getVersionPacket(mode: Mode) []const u8 {
+                const write_len = @min(data.len, std.math.maxInt(i32));
+
+                // create write request
+                const write_req_slice = bun.default_allocator.dupe(u8, data[0..write_len]) catch bun.outOfMemory();
+                const write_req = bun.new(WindowsWrite, .{
+                    .owner = this,
+                    .write_slice = write_req_slice,
+                    .write_req = std.mem.zeroes(uv.uv_write_t),
+                    .write_buffer = uv.uv_buf_t.init(write_req_slice),
+                });
+                bun.assert(this.windows.windows_write == null);
+                this.windows.windows_write = write_req;
+
+                pipe.ref(); // ref on write
+                if (this.windows.windows_write.?.write_req.write(pipe.asStream(), &this.windows.windows_write.?.write_buffer, write_req, &_windowsOnWriteComplete).asErr()) |err| {
+                    _windowsOnWriteComplete(write_req, @enumFromInt(-@as(c_int, err.errno)));
+                }
+                // write request is queued. it will call _onWriteComplete when it completes.
+            },
+            false => {
+                if (fd) |fd_unwrapped| {
+                    this._onWriteComplete(socket.writeFd(data, fd_unwrapped));
+                } else {
+                    this._onWriteComplete(socket.write(data, false));
+                }
+            },
+        };
+    }
+    fn _windowsOnWriteComplete(write_req: *WindowsWrite, status: uv.ReturnCode) void {
+        log("SendQueue#_windowsOnWriteComplete", .{});
+        const write_len = write_req.write_slice.len;
+        const this = blk: {
+            defer write_req.destroy();
+            break :blk write_req.owner orelse return; // orelse case if disconnected before the write completes
+        };
+
+        const vm = JSC.VirtualMachine.get();
+        vm.eventLoop().enter();
+        defer vm.eventLoop().exit();
+
+        this.windows.windows_write = null;
+        if (this.getSocket()) |socket| socket.unref(); // write complete; unref
+        if (status.toError(.write)) |_| {
+            this._onWriteComplete(-1);
         } else {
-            this.closeTask();
+            this._onWriteComplete(@intCast(write_len));
         }
-    }
 
-    pub fn closeTask(this: *NamedPipeIPCData) void {
-        log("NamedPipeIPCData#closeTask is_server {}", .{this.is_server});
-        if (this.disconnected) {
-            _ = this.writer.flush();
-            this.writer.end();
-            if (this.writer.getStream()) |stream| {
-                stream.readStop();
-            }
-            if (!this.writer.hasPendingData()) {
-                this.detach();
-            }
+        if (this.windows.try_close_after_write) {
+            this.closeSocket(.normal, .user);
         }
     }
+    fn getGlobalThis(this: *SendQueue) ?*JSC.JSGlobalObject {
+        return switch (this.owner) {
+            inline else => |owner| owner.globalThis,
```

Comment:
globalThis doesn't need to be optional in `IPCInstance`. We can change this function to return a non-optional globalThis pointer